[cmake] iconv is a required dependency

bkuhls opened the pull request:
This patch adds support for libiconv on non-WIN32 platforms  currently
only provided by the autoconf-based build system:
https://github.com/xbmc/xbmc/blob/Krypton/configure.ac#L1172

This commit fixes an error during linking with an uClibc-based
buildroot toolchain:

[100%] Linking CXX executable kodi.bin
/home/buildroot/br8_ffmpeg3_kodi17_github/output/host/usr/lib/gcc/i586-buildroot-linux-uclibc/6.3.0/../../../../i586-buildroot-linux-uclibc/bin/ld: build/utils/utils.a(CharsetConverter.cpp.o): undefined reference to symbol 'libiconv_open'
/home/buildroot/br8_ffmpeg3_kodi17_github/output/host/usr/i586-buildroot-linux-uclibc/sysroot/usr/lib32/libiconv.so.2: error adding symbols: DSO missing from command line

Signed-off-by: Bernd Kuhls <bernd.kuhls@t-online.de>

<!--- Provide a general summary of your change in the Title above -->

## Description
<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

wsnipex commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,138 @@
+)
+
+if(NOT ICONV_INCLUDE_DIR STREQUAL "ICONV_INCLUDE_DIR-NOTFOUND")
+    set(CMAKE_REQUIRED_INCLUDES ${ICONV_INCLUDE_DIR})
```
duplicated further down, so can be removed here

wsnipex commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,138 @@
+
+if(MSVC_VERSION GREATER 1800)
+    set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} legacy_stdio_definitions.lib)
+endif()
```
since we call this module only for non-windows, this block is superfluous

wsnipex commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,138 @@
+if(ICONV_FOUND)
+  set(ICONV_LIBRARIES ${ICONV_LIBRARY})
+else(ICONV_FOUND)
+  set(ICONV_LIBRARIES)
```
not needed

wsnipex commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,138 @@
+  set(ICONV_LIBRARIES)
+endif(ICONV_FOUND)
+
+if(ICONV_FOUND)
```
please merge with the if block above

wsnipex commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,138 @@
+
+    if(MSVC_VERSION GREATER 1800)
+        set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} legacy_stdio_definitions.lib)
+    endif()
```
again, no win32 stuff needed

wsnipex reviewed (changes requested):
thanks. Only a few minor cleanups please.

bkuhls commented:
@wsnipex : done

wsnipex commented:
jenkins build this please

bkuhls commented:
@wsnipex: updated patch to fix rbpi build error

wsnipex commented:
I'm not sure we should skip the check on rpbi, @popcornmix?
jenkins build this please

popcornmix commented:
I would have thought rbpi should behave the same as linux so should be enabled.
What is the problem? Does this not handle cross-compiles?

bkuhls commented:
@wsnipex @popcornmix: The previous patch failed on rbpi: http://jenkins.kodi.tv/job/LINUX-RBPI/9535/console

MartijnKaijser commented:
ping

bkuhls commented:
Hi, I tried to reproduce the build error caught by jenkins in rbpi but I am having problems to find the correct instructions for setting up the cross-compile env. When adding the patch to Libreelec master it works when build with "PROJECT=RPi DEVICE=RPi2 ARCH=arm make":
```
-- Core system type: rbpi
-- Platform: 
-- CPU: arm, ARCH: arm
-- Cross-Compiling: TRUE
[...]
-- Found ICONV: In glibc  
-- One (and only one) of the ICONV_ACCEPTS_... tests must pass
```

Rechi commented on CMakeLists.txt:
```diff
@@ -128,6 +128,9 @@ endif()
 if(CORE_SYSTEM_NAME STREQUAL android)
   list(APPEND required_deps Zip)
 endif()
+if(NOT WIN32 AND NOT CORE_SYSTEM_NAME STREQUAL rbpi)
```
`... AND NOT CORE_PLATFORM_NAME_LC STREQUAL rbpi`

bkuhls replied:
fixed, thx!

Rechi commented:
Was there already a final decision if finding iconv on rbpi can be skipped?
Windows also needs iconv so why is the check skipped for it too?

bkuhls commented:
Let's try a different approach, I found a FindIconv.cmake file which looks more promising w.r.t. cross-platform compatibility. Unfortunately I can not test win/mac builds myself so I have to wait for Jenkins results, this may take several iterations in the worst case. Cross-compiling for "Core system type: linux" with glibc and uClibc using buildroot worked however.

wsnipex commented:
this FindIconv is worse then the old one.

Rechi commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,43 @@
+#  ICONV_LIBRARIES - Link these to use Iconv
+#  ICONV_SECOND_ARGUMENT_IS_CONST - the second argument for iconv() is const
+#
+include(CheckCXXSourceCompiles)
```
not needed

Rechi commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,43 @@
+
+IF (ICONV_INCLUDE_DIR AND ICONV_LIBRARIES)
+  # Already in cache, be silent
+  SET(ICONV_FIND_QUIETLY TRUE)
```
why is this needed

Rechi commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,43 @@
+FIND_LIBRARY(ICONV_LIBRARIES NAMES iconv libiconv libiconv-2 c)
+
+IF(ICONV_INCLUDE_DIR AND ICONV_LIBRARIES)
+  SET(ICONV_FOUND TRUE)
```
this has to be done with `find_package_handle_standard_args`

Rechi commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,43 @@
+ENDIF(ICONV_INCLUDE_DIR AND ICONV_LIBRARIES)
+
+set(CMAKE_REQUIRED_INCLUDES ${ICONV_INCLUDE_DIR})
+set(CMAKE_REQUIRED_LIBRARIES ${ICONV_LIBRARIES})
```
why is this needed?

Rechi commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,43 @@
+
+IF(ICONV_FOUND)
+  IF(NOT ICONV_FIND_QUIETLY)
+    MESSAGE(STATUS "Found Iconv: ${ICONV_LIBRARIES}")
```
`find_package_handle_standard_args` handles this whole logic

Rechi reviewed (changes requested):
This doesn't fit the style we use for the other find modules.

bkuhls commented:
Pushed another try with a simplified FindIconv.cmake based on the review by @Rechi.

I moved Iconv to optional_deps now because
- the current situation assumes that iconv functions are provided by libc, which is used for linking anyway
- my patch tries to solve a link error for (those rare) toolchains where libc does not provide iconv functions
- properly checking for Iconv using cmake is not easy (cmake itself does not provide a module for this purpose (an upstream bug ( https://cmake.org/Bug/view.php?id=13517 ), dating back to 2012, was closed last year ( https://gitlab.kitware.com/cmake/cmake/issues/13517 ) without an solution, the other FindIconv.cmake files I tested here did not work for all platforms
- when libiconv is found we can assume we should use it, if it was not found we should not throw an error because building Kodi worked without it with the exception of toolchains lacking iconv functions (buildroot only allows the usage of libiconv when the toolchain does not provide locale functions itself: https://github.com/buildroot/buildroot/blob/master/package/libiconv/Config.in#L3 ). This is why I moved Iconv to optional_deps and hope that Jenkins is happy with it ;)

bkuhls commented:
Analysis of Jenkins errors:

Win-64, LINUX-64 & IOS-ARM64: did not even start building because of a git error

LINUX-RBPI: did Jenkins use the correct patch? Iconv is not part of core_require_dep anymore
```  Could not find Iconv
Call Stack (most recent call first):
  cmake/scripts/common/Macros.cmake:340 (find_package)
  CMakeLists.txt:167 (core_require_dep)```

Rechi commented:
> I moved Iconv to optional_deps now [...]

This is wrong, iconv is a hard requirement for all platforms.

bkuhls commented:
Ok, next try: This time FindIconv.cmake is meant to find the function iconv_open as hard dependency, not libiconv.
First we check for function iconv_open with the standard set of libraries, this means we assume that iconv functions are provided by libc. Only if they are not found we check for an iconv library containing iconv_open
```-- Found Iconv: /home/buildroot/br8/output/host/x86_64-buildroot-linux-uclibc/sysroot/usr/lib/libiconv.so```

If all of this fails:
```CMake Error at /usr/share/cmake-3.7/Modules/FindPackageHandleStandardArgs.cmake:138 (message):
  Could NOT find Iconv (missing: ICONV_LIBRARY)
Call Stack (most recent call first):
  /usr/share/cmake-3.7/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  modules/FindIconv.cmake:35 (find_package_handle_standard_args)
  scripts/common/Macros.cmake:354 (find_package)
  CMakeLists.txt:140 (core_require_dep)```

wsnipex commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,40 @@
+  if(ICONV)
+    set(ICONV_INCLUDE_DIRS ${ICONV_INCLUDE_DIR})
+    set(ICONV_LIBRARIES ${ICONV_LIBRARY})
+  endif()
```
please move this block down after find_package_handle_standard_args, but make it if(ICONV_FOUND)

wsnipex commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,40 @@
+
+  include(FindPackageHandleStandardArgs)
+  find_package_handle_standard_args(Iconv
+                                  REQUIRED_VARS ICONV_LIBRARY ICONV_INCLUDE_DIR)
```
and add ICONV here. It might also make sense to rename Iconv to something not possibly clashing with the upper case var ICONV. In that case adjust the ICONV_FOUND block mentioned in my comment above.

wsnipex reviewed (dismissed):
looks much better, just a minor change please

bkuhls commented:
Pushed new version, tested with buildroot git master:
```
Krypton - uclibc
-- Found Iconv: /home/buildroot/br8_ffmpeg3_kodi17_github/output/host/i686-buildroot-linux-uclibc/sysroot/usr/lib/libiconv.so  

Krypton - glibc
-- Found Iconv: 1  

Leia - uclibc
-- Found Iconv: /home/buildroot/br8_ffmpeg3_kodi17_github/output/host/i686-buildroot-linux-uclibc/sysroot/usr/lib/libiconv.so  

Leia - glibc
-- Found Iconv: 1  ```

Rechi commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,36 @@
+  find_package_handle_standard_args(Iconv
+                                    REQUIRED_VARS ICONV_LIBRARY HAVE_ICONV_IN_LIBICONV)
+
+  if(HAVE_ICONV_IN_LIBICONV)
```
`if(ICONV_FOUND)`
`ICONV_FOUND` gets set by `find_package_handle_standard_args`

bkuhls replied:
ok

wsnipex commented:
jenkins build this please

Rechi reviewed (changes requested):
`find_path` for iconv.h is now missing for the non HAVE_ICONV_IN_LIBC case
everything else looks good

bkuhls commented:
In the next iteration of this patch I removed
```
set(CMAKE_REQUIRED_INCLUDES iconv.h)
```
so I think the find_path for iconv.h is not needed anymore, check_library_exists uses "iconv.h", no additional path needed imho.

wsnipex commented on cmake/modules/FindIconv.cmake:
```diff
@@ -0,0 +1,35 @@
+
+include(FindPackageHandleStandardArgs)
+
+check_library_exists("" iconv_open "iconv.h" HAVE_ICONV_IN_LIBC)
```
this doesn't work like you think. LOCATION is intended to be a directory, not a header file.
see https://github.com/Kitware/CMake/blob/master/Modules/CheckLibraryExists.cmake#L61

bkuhls commented:
@wsnipex: Done.
Sorry for using again a totally new FindIconv.cmake file, I never coded for CMake in the past, so I have to search the internet for ideas. While searching for the correct syntax for check_library_exists I came across https://github.com/owncloud/client/blob/fe0de111fdd4eef1c803cc3aca13d4f70c12b7ce/cmake/modules/FindIconv.cmake which looks as it contains everything needed for our purpose.
Let's see if Jenkins is happy with it ;)

wsnipex commented:
I'm not going to bother reviewing anymore, since you keep ignoring it and replace the file completely instead.

Rechi reviewed (changes requested):
This is getting too complex again.
Restore the previous version of your patch and just do the requested changes.

Rechi commented:
@bkuhls you still plan to do the requested changes?

bkuhls commented:
Yes, but after PR 12942 landed, which added FindIConv.cmake, I have to rework this PR.

bkuhls commented:
Done, it seems the new FindIConv.cmake also works fine with uClibc, only required_deps needed to be changed. Tested with glibc & uClibc.

Rechi commented:
One last thing, iconv has to be removed from
https://github.com/xbmc/xbmc/blob/5837475d/cmake/platform/windowsstore/windowsstore.cmake

bkuhls commented:
@Rechi: Done.
I just tested a backport of this PR together with FindIConv.cmake from PR 12942 successfully with Krypton HEAD. If this PR is accepted does it make sense that I send an update to PR 11665?

Rechi commented:
wrong revision was build
jenkins build this please

wsnipex commented:
the findiconv modules from the UWP pull is just as broken.
It fails when iconv is built-in:

-- Could NOT find iconv (missing:  ICONV_LIBRARY)

Rechi committed:
iconv is a required dependency

Rechi commented:
@bkuhls is https://github.com/xbmc/xbmc/commit/a78fc281229c62f76c85f61997fb30d50a3cd016 working for you?

bkuhls commented:
@Rechi: Yes, Kodi git master with your patch still works fine with libiconv:
```
-- Found Iconv: /home/buildroot/br8/output/host/x86_64-buildroot-linux-uclibc/sysroot/usr/lib/libiconv.so
[...]
$ output/host/bin/x86_64-buildroot-linux-uclibc-readelf -a  output/target/usr/lib/kodi/kodi.bin | grep NEEDED | grep iconv
 0x0000000000000001 (NEEDED)             Shared library: [libiconv.so.2]
$ output/host/bin/x86_64-buildroot-linux-uclibc-g++ -v
[...]
gcc-Version 6.4.0 (Buildroot 2017.11-git-00994-ge467773c29-dirty) 
```

Rechi commented:
@bkuhls please update this PR with my patch

Rechi commented:
wrong revision was build
jenkins build this please

bkuhls commented:
@Rechi: The same problem I saw back in February ( https://github.com/xbmc/xbmc/pull/11665#issuecomment-279990639 ) again occurs with your patch on rbpi. We should treat libiconv as an optional dependency and do not bother to know which target library contains iconv functions if libiconv is not present.
@wsnipex: Regarding https://github.com/xbmc/xbmc/pull/11665#issuecomment-341655778 I would like to point out that this message is just an information that libiconv could not be found, the resulting build worked fine.

Rechi commented:
@bkuhls please leaf the PR as it currently is.
@wsnipex and I found the real reason why rbpi currently fails. A fix is very likely coming soon.

wsnipex commented:
jenkins build this please

wsnipex commented:
jenkins build this please

bkuhls commented:
Considering the fact that this PR needed changes for the rbpi toolchain do you think a backport of this PR to Krypton will be accepted?

Rechi commented:
The rbpi toolchain change won't be backported, so this also can't be backported as it would break rbpi.
